hugegraph-pd/hg-pd-client/src/main/java/org/apache/hugegraph/pd/client/ClientCache.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hugegraph.pd.client;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.hugegraph.pd.common.GraphCache;
+import org.apache.hugegraph.pd.common.KVPair;
+import org.apache.hugegraph.pd.common.PDException;
+import org.apache.hugegraph.pd.common.PartitionUtils;
+import org.apache.hugegraph.pd.grpc.Metapb;
+import org.apache.hugegraph.pd.grpc.Metapb.Partition;
+import org.apache.hugegraph.pd.grpc.Metapb.Shard;
+import org.apache.hugegraph.pd.grpc.Metapb.ShardGroup;
+import org.apache.hugegraph.pd.grpc.Pdpb.CachePartitionResponse;
+import org.apache.hugegraph.pd.grpc.Pdpb.CacheResponse;
+
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ClientCache {
+
+    private final AtomicBoolean initialized = new AtomicBoolean(false);
+    private final org.apache.hugegraph.pd.client.PDClient client;
+    private volatile Map<Integer, KVPair<ShardGroup, Shard>> groups;
+    private volatile Map<Long, Metapb.Store> stores;
+    private volatile Map<String, GraphCache> caches = new ConcurrentHashMap<>();
+
+    public ClientCache(org.apache.hugegraph.pd.client.PDClient pdClient) {
+        groups = new ConcurrentHashMap<>();
+        stores = new ConcurrentHashMap<>();
+        client = pdClient;
+    }
```

Comment:
Maybe we can consider to put the initialization of member variables all in the constructor for a cleaner code stype. This can be fixed in the next stype related PR. @VGalaxies

Reply:
> Maybe we can consider to put the initialization of member variables all in the constructor for a cleaner code style. This can be fixed in the next style related PR. @VGalaxies

Yep, we could mark a `TODO` first & enhance it with the v4.0 together